pkg/controllers/util/listwatcher/listwatcher.go

```diff
@@ -0,0 +1,239 @@
+// is processed immediately to allow the allocated memory to be potentially garbage collected before the next request.
+func NewPagedListWatcher(lw cache.ListerWatcher, pageSize int64) ListWatcher {
+	return &pagedListWatcher{
+		lw:        lw,
+		receivers: []chan<- Event{},
+		pageSize:  pageSize,
+		mu:        sync.Mutex{},
+		mux:       make(chan Event, 100),
+		backoff:   wait.NewExponentialBackoffManager(800*time.Millisecond, 30*time.Second, 2*time.Minute, 2.0, 1.0, &clock.RealClock{}),
+	}
+}
+
+func (p *pagedListWatcher) AddReceiver(receiver chan<- Event) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	p.receivers = append(p.receivers, receiver)
+}
+
+func (p *pagedListWatcher) Start(ctx context.Context) {
+	logger := klog.FromContext(ctx).WithValues("component", "paged-list-watcher")
+	listCh := make(chan string, 1)
+
+	go func() {
+		for {
+			// list resource
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				p.mux <- Event{Type: StartSync}
+
+				resourceVersion, err := p.list(ctx)
+				if err != nil {
+					logger.Error(err, "Failed to list resource, will retry after backoff")
+					<-p.backoff.Backoff().C()
+					continue
+				}
+
+				p.mux <- Event{Type: EndSync}
+				listCh <- resourceVersion
+			}
+
+			// watch resource
+			select {
+			case <-ctx.Done():
+				return
+			case resourceVersion := <-listCh:
+				if err := p.watch(ctx, resourceVersion); err != nil {
+					logger.Error(err, "Watch resource ended with error, will relist after backoff")
+					<-p.backoff.Backoff().C()
+				} else {
+					logger.V(4).Info("Watch resource ended gracefully, will relist")
+				}
+			}
+		}
+	}()
+
+	go func() {
+		for event := range p.mux {
+			p.mu.Lock()
+			for _, receiver := range p.receivers {
+				receiver <- event
+			}
+			p.mu.Unlock()
+		}
+	}()
+}
+
+func (p *pagedListWatcher) list(ctx context.Context) (rv string, err error) {
+	var resourceVersion string
+	listOpts := metav1.ListOptions{
+		Limit:    p.pageSize,
+		Continue: "",
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return "", nil
+		default:
+			resp, err := p.lw.List(listOpts)
+			if err != nil {
+				return "", fmt.Errorf("failed to list resource: %w", err)
+			}
+
+			list, err := meta.ListAccessor(resp)
+			if err != nil {
+				return "", fmt.Errorf("failed to list resource: %w", err)
+			}
+
+			listOpts.Continue = list.GetContinue()
+			resourceVersion = list.GetResourceVersion()
+
+			err = meta.EachListItem(resp, func(o runtime.Object) error {
+				p.mux <- Event{
+					Type:   Add,
+					Object: o.DeepCopyObject(),
+				}
+				return nil
+			})
+
+			if err != nil {
+				return "", fmt.Errorf("failed to process list item: %w", err)
+			}
+		}
+
+		if len(listOpts.Continue) == 0 {
+			break
+		}
+	}
+
+	return resourceVersion, nil
+}
+
+func (p *pagedListWatcher) watch(ctx context.Context, resourceVersion string) error {
+	watcher, err := p.lw.Watch(metav1.ListOptions{ResourceVersion: resourceVersion})
```

Comment:
https://github.com/kubernetes/client-go/blob/8cbca742aebe24b24f7f4e32fd999942fa9133e8/tools/cache/reflector.go#L302

Do we need `TimeoutSeconds` and `AllowWatchBookmarks`?

```go
		options := metav1.ListOptions{
			ResourceVersion: r.LastSyncResourceVersion(),
			// We want to avoid situations of hanging watchers. Stop any watchers that do not
			// receive any events within the timeout window.
			TimeoutSeconds: &timeoutSeconds,
			// To reduce load on kube-apiserver on watch restarts, you may enable watch bookmarks.
			// Reflector doesn't assume bookmarks are returned at all (if the server do not support
			// watch bookmarks, it will ignore this field).
			AllowWatchBookmarks: true,
		}
```